File: boilerplate/nestjs/order-service/src/infrastructure/api/auth.controller.ts
import {
  Controller,
  Post,
  Get,
  Body,
  HttpCode,
  HttpStatus,
  Inject,
  Headers,
  UnauthorizedException,
  Res,
} from "@nestjs/common";
import { Response } from "express";

import { LoginCommand, LoginResult } from "../../application/dtos/auth.dto";
import {
  RefreshTokenCommand,
  RefreshTokenResult,
} from "../../application/dtos/refresh-token.dto";
import {
  RegisterCommand,
  RegisterResult,
} from "../../application/dtos/user-order.dto";
import { UserProfileResult } from "../../application/dtos/user-order.dto";
import { IAuthenticateUserUseCase } from "../../application/usecases/authenticate-user.use-case.interface";
import { IGetCurrentUserUseCase } from "../../application/usecases/get-current-user.use-case.interface";
import { ILogoutUseCase } from "../../application/usecases/logout.use-case.interface";
import { IRefreshTokenUseCase } from "../../application/usecases/refresh-token.use-case.interface";
import { IRegisterUserUseCase } from "../../application/usecases/register-user.use-case.interface";
import { ITokenParser } from "../../domain/ports/token-parser.port";

@Controller("api/v1/auth")
export class AuthController {
  constructor(
    @Inject("IAuthenticateUserUseCase")
    private readonly authenticateUserUseCase: IAuthenticateUserUseCase,
    @Inject("IRegisterUserUseCase")
    private readonly registerUserUseCase: IRegisterUserUseCase,
    @Inject("IGetCurrentUserUseCase")
    private readonly getCurrentUserUseCase: IGetCurrentUserUseCase,
    @Inject("IRefreshTokenUseCase")
    private readonly refreshTokenUseCase: IRefreshTokenUseCase,
    @Inject("ILogoutUseCase")
    private readonly logoutUseCase: ILogoutUseCase,
    @Inject("ITokenParser")
    private readonly tokenParser: ITokenParser,
  ) {}

  @Post("register")
  @HttpCode(HttpStatus.CREATED)
  async register(@Body() command: RegisterCommand): Promise<RegisterResult> {
    return this.registerUserUseCase.execute(command);
  }

  @Post("login")
  @HttpCode(HttpStatus.OK)
  async login(
    @Body() command: LoginCommand,
    @Res({ passthrough: true }) response: Response,
  ): Promise<LoginResult> {
    const result = await this.authenticateUserUseCase.execute(command);
    // Set httpOnly cookies for token storage (XSS prevention)
    response.cookie("access_token", result.accessToken, {
      httpOnly: true,
      secure: process.env.NODE_ENV === "production",
      sameSite: "strict",
      maxAge: 3600 * 1000, // 1 hour
    });
    response.cookie("refresh_token", result.refreshToken, {
      httpOnly: true,
      secure: process.env.NODE_ENV === "production",
      sameSite: "strict",
      maxAge: 24 * 3600 * 1000, // 24 hours
    });
    return result;
  }

  @Post("refresh")
  @HttpCode(HttpStatus.OK)
  async refresh(
    @Body() command: RefreshTokenCommand,
    @Res({ passthrough: true }) response: Response,
  ): Promise<RefreshTokenResult> {
    const result = await this.refreshTokenUseCase.execute(command);
    // Rotate cookies with new token pair
    response.cookie("access_token", result.accessToken, {
      httpOnly: true,
      secure: process.env.NODE_ENV === "production",
      sameSite: "strict",
      maxAge: 3600 * 1000,
    });
    response.cookie("refresh_token", result.refreshToken, {
      httpOnly: true,
      secure: process.env.NODE_ENV === "production",
      sameSite: "strict",
      maxAge: 24 * 3600 * 1000,
    });
    return result;
  }

  @Post("logout")
  @HttpCode(HttpStatus.NO_CONTENT)
  async logout(
    @Headers("authorization") authHeader: string,
    @Res({ passthrough: true }) response: Response,
  ): Promise<void> {
    if (authHeader && authHeader.startsWith("Bearer ")) {
      const token = authHeader.split(" ")[1];
      const userId = this.tokenParser.parseUserId(token);
      if (userId) {
        await this.logoutUseCase.execute(userId);
      }
    }
    response.clearCookie("access_token");
    response.clearCookie("refresh_token");
  }

  @Get("me")
  @HttpCode(HttpStatus.OK)
  async me(
    @Headers("authorization") authHeader: string,
  ): Promise<UserProfileResult> {
    if (!authHeader || !authHeader.startsWith("Bearer ")) {
      throw new UnauthorizedException("Authentication required");
    }
    const token = authHeader.split(" ")[1];
    const userId = this.tokenParser.parseUserId(token);
    if (!userId) {
      throw new UnauthorizedException("Invalid token");
    }
    return this.getCurrentUserUseCase.execute(userId);
  }
}

File: boilerplate/python/order-service/src/infrastructure/api/auth_controller.py
from __future__ import annotations
from fastapi import APIRouter, Depends, HTTPException, status, Response
from fastapi.security import HTTPBearer, HTTPAuthorizationCredentials
from .dependencies import get_authenticate_user_use_case, get_register_user_use_case, get_change_password_use_case, get_current_user_use_case, get_token_parser
from application.dtos import LoginCommand, LoginResult, RegisterCommand, RegisterResult, ChangePasswordCommand, UserProfileResult
from domain.models.user import UserId

router = APIRouter(prefix="/auth", tags=["Authentication"])
security = HTTPBearer()

@router.post("/register", response_model=RegisterResult, status_code=status.HTTP_201_CREATED)
async def register(command: RegisterCommand, use_case=Depends(get_register_user_use_case)):
    return use_case.execute(command)

@router.post("/login", response_model=LoginResult)
async def login(command: LoginCommand, response: Response, use_case=Depends(get_authenticate_user_use_case)):
    result = use_case.execute(command)
    # Set httpOnly cookies for token storage (security: XSS prevention)
    response.set_cookie(
        key="access_token",
        value=result.accessToken,
        httponly=True,
        secure=False,  # Set to True in production (HTTPS)
        samesite="strict",
        max_age=3600,
    )
    response.set_cookie(
        key="refresh_token",
        value=result.refreshToken,
        httponly=True,
        secure=False,  # Set to True in production (HTTPS)
        samesite="strict",
        max_age=86400,
    )
    return result

@router.post("/refresh", response_model=LoginResult)
async def refresh(
    response: Response,
    credentials: HTTPAuthorizationCredentials = Depends(security),
    token_parser=Depends(get_token_parser),
):
    """Refresh access token using a valid refresh token."""
    user_id = token_parser.parse_user_id(credentials.credentials)
    if not user_id:
        raise HTTPException(status_code=status.HTTP_401_UNAUTHORIZED, detail="Invalid or expired refresh token")

    # TODO: Implement proper refresh token rotation (new token pair generation)
    # For now, just validate the token exists and return success
    # In production, generate new_access_token + new_refresh_token here

    return {"message": "Token refresh not fully implemented — see auth-flow.md Phase 2"}

@router.post("/logout")
async def logout(
    response: Response,
    credentials: HTTPAuthorizationCredentials = Depends(security),
    token_parser=Depends(get_token_parser),
):
    """Logout and invalidate tokens."""
    # TODO: Add token to Redis blacklist with TTL matching token expiry
    # For now, just clear cookies
    user_id = token_parser.parse_user_id(credentials.credentials)
    if user_id:
        pass  # TODO: Blacklist token

    response.delete_cookie("access_token")
    response.delete_cookie("refresh_token")
    return {"message": "Logged out successfully"}

@router.get("/me", response_model=UserProfileResult)
async def me(
    credentials: HTTPAuthorizationCredentials = Depends(security),
    use_case=Depends(get_current_user_use_case),
    token_parser=Depends(get_token_parser)
):
    user_id = token_parser.parse_user_id(credentials.credentials)
    if not user_id:
        raise HTTPException(status_code=status.HTTP_401_UNAUTHORIZED, detail="Invalid or expired token")
    return use_case.execute(user_id)

@router.post("/change-password")
async def change_password(
    command: ChangePasswordCommand,
    credentials: HTTPAuthorizationCredentials = Depends(security),
    use_case=Depends(get_change_password_use_case),
    token_parser=Depends(get_token_parser)
):
    user_id = token_parser.parse_user_id(credentials.credentials)
    if not user_id:
        raise HTTPException(status_code=status.HTTP_401_UNAUTHORIZED, detail="Invalid or expired token")
    use_case.execute(str(user_id.value), command)
    return {"message": "Password changed successfully"}
